### PR TITLE
port apache mynewt

### PR DESCRIPTION
**Pull request by Naruto**
# how to build
### wolfssl
fetch working branch for porting mynewt.

```
git remote add naruto git@github.com:Naruto/wolfssl.git
git fetch naruto
git checkout feature/mynewt
```

### mynewt project
create new mynewt project. It is targeted wolfssl source codes.

```
newt new newproject # create mynewt project
cd newproject
newt upgrade
```

## deploy wolfssl to mynewt project
deploy wolfssl sources to mynewt project

```
cd /path/to/wolfssl
./IDE/mynewt/setup.sh /path/to/newproject
```

### CODE REVIEW

File: src/wolfio.c [hunk 1]
```diff
@@ -1850,4 +1850,227 @@ int MicriumGenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *ctx)
...
+/* return Mynewt_Ctx instance */
+void* mynewt_ctx_new() {
+    int rc = 0;
+    Mynewt_Ctx *mynewt_ctx = NULL;
+
+    mynewt_ctx = (Mynewt_Ctx *)XMALLOC(sizeof(struct Mynewt_Ctx), NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    if(!mynewt_ctx) return NULL;
+
+    XMEMSET(mynewt_ctx, 0, sizeof(Mynewt_Ctx));
+    mynewt_ctx->mnMemBuffer = XMALLOC(OS_MEMPOOL_BYTES(MYNEWT_VAL(WOLFSSL_MNSOCK_MEM_BUF_COUNT), MYNEWT_VAL(WOLFSSL_MNSOCK_MEM_BUF_SIZE)), 0, 0);
```
Comment (cconlon) [inline src/wolfio.c]: Our coding standard requires us to limit line lengths to 80 characters.  Can you adjust your code to wrap any lines that are longer than that?  Thanks!

File: wolfssl/wolfcrypt/settings.h [hunk 2]
```diff
@@ -1196,6 +1199,44 @@ extern void uITRON4_free(void *p) ;
...
+    #define NO_DEV_RANDOM
+    #define NO_DH
+    #define NO_WOLFSSL_DIR
+    #define NO_ERROR_STRINGS
+    #define HAVE_ECC
+    #define NO_SESSION_CACHE
+    #define USE_SLOW_SHA
+    #define NO_WOLFSSL_SERVER
+    #define NO_ERROR_STRINGS
+    // #define WC_NO_HARDEN // TODO:
```
[the remark this answers is not in the document]
Comment (Naruto) [reply wolfssl/wolfcrypt/settings.h]: I will remove.

File: IDE/mynewt/apps.wolfsslclienttlsmn.syscfg.yml [hunk 3]
```diff
@@ -0,0 +1,42 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
```
[the remark this answers is not in the document]
Comment (cconlon) [reply IDE/mynewt/apps.wolfsslclienttlsmn.syscfg.yml]: It looks like this file still includes an Apache license.

[the remark this answers is not in the document]
Comment (Naruto) [reply IDE/mynewt/apps.wolfsslclienttlsmn.syscfg.yml]: sorry 🙇
I will update

File: src/wolfio.c [hunk 4]
```diff
@@ -1850,4 +1850,220 @@ int MicriumGenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *ctx)
...
+        goto done;
+    }
+    rc = mn_sendto(mynewt_ctx->mnSocket, m, (struct mn_sockaddr *)&mynewt_ctx->mnSockAddrIn);
+    if(rc != 0) {
+        WOLFSSL_MSG("Mynewt Send mn_sendto error");
+        goto done;
+    }
+    write_sz = sz;
+
+done:
```
[the remark this answers is not in the document]
Comment (Naruto) [reply src/wolfio.c]: Yes, i will.

File: wolfssl/wolfcrypt/settings.h [hunk 5]
```diff
@@ -1196,6 +1199,43 @@ extern void uITRON4_free(void *p) ;
...
+    #define NO_WRITEV
+    #define WOLFSSL_USER_IO
+    #define SINGLE_THREADED
+    #define NO_DEV_RANDOM
+    #define NO_DH
+    #define NO_WOLFSSL_DIR
+    #define NO_ERROR_STRINGS
+    #define HAVE_ECC
+    #define NO_SESSION_CACHE
+    #define USE_SLOW_SHA
```
[the remark this answers is not in the document]
Comment (Naruto) [reply wolfssl/wolfcrypt/settings.h]: No.
I will remove

File: IDE/mynewt/client-tls-mn.c [hunk 6]
```diff
@@ -0,0 +1,628 @@
...
+#include <log/log.h>
+#include <config/config.h>
+#include <hal/hal_system.h>
+
+#include <bootutil/image.h>
+#include <bootutil/bootutil.h>
+
+#include <shell/shell.h>
+#include <mn_socket/mn_socket.h>
+// #include <inet_def_service/inet_def_service.h>
```
[the remark this answers is not in the document]
Comment (Naruto) [reply IDE/mynewt/client-tls-mn.c]: OK.
I will move client-tls-mn.c to wolfssl-examples repository, and update `IDE/mynewt/setup.sh` as pulls wolfssl-examples repository after finish to review.

File: IDE/mynewt/client-tls-mn.c [hunk 7]
```diff
@@ -0,0 +1,628 @@
...
+        return -1;
+    }
+
+    int ret = wolfSSL_CTX_load_verify_buffer(wolfsslCtx, ca_cert_der_2048, sizeof_ca_cert_der_2048, SSL_FILETYPE_ASN1);
+    if(ret != SSL_SUCCESS) {
+        console_printf("Error %d loading CA cert\n", ret);
+        return ret;
+    }
+
+    wolfSSL_CTX_set_verify(wolfsslCtx, SSL_VERIFY_NONE, 0);
```
[the remark this answers is not in the document]
Comment (Naruto) [reply IDE/mynewt/client-tls-mn.c]: connect to `www.example.com`

**Comment by cconlon** [comment]
Thanks for addressing my requested changes.  Can you squash the commits in your pull request down into a smaller number of commits with meaningful commit messages?  This will make our master branch cleaner once we merge your PR.  Thanks!

**Comment by cconlon** [comment]
Overall, your instructions worked well for me, thanks!

When I tried to run the install script, I got the following error:

```
./IDE/mynewt/setup.sh: 25: ./IDE/mynewt/setup.sh: pushd: not found
```

I got around this temporarily by changing the script to use "#!/bin/bash" instead of "#!/bin/sh". I think we want to leave this as a shell script instead of a bash script for portability. Can you look into a solution for this?

**Comment by cconlon** [comment]
Thanks!  I'll wait to merge these until changes are made to the example instructions in 'wolfssl-examples':

https://github.com/wolfSSL/wolfssl-examples/pull/100

**Comment by cconlon** [comment]
Naruto-san,

Thanks for updating the example/sample README in https://github.com/wolfSSL/wolfssl-examples/pull/100.  I tested this out on a Ubuntu 17.10 VM and it worked successfully for me.

Two more (and hopefully final) requests before we can merge this pull request:

1.  Please rebase this on top of the master branch to resolve merge conflicts.

2.  All new files added to the repository need to be listed in an include.am file so that they are included in our distribution bundle when we run "make dist".  A few notes on how to do this:

- Please create a new include.am file inside the "IDE/mynewt" directory, similar to the ones in other "IDE/*" directories.  "IDE/mynewt/include.am" should add the files in that directory to EXTRA_DIST.
- After creating "IDE/mynewt/include.am", add this as an include in "IDE/include.am".

Thanks!
Chris

**Comment by Naruto** [comment]
I've added `IDE/mynewt/include.am`, `IDE/include.am`  and squashed commit. 
Please review it.

**Comment by cconlon** [comment]
Retest this please Jenkins

**Pull request merged**